README.md
fastcache
=========

C implementation of Python 3 lru_cache for Python 2.6, 2.7, 3.2, 3.3, 3.4

Passes all tests in the standard library for functools.lru_cache.

Obeys same API as Python 3.3/3.4 functools.lru_cache with 2 enhancements:

1.  An additional argument `state` may be supplied which must be a `list` or `dict`.  This allows one to safely cache functions for which the result depends on some context which is not a part of the function call signature.
2.  An additional argument `unhashable` may be supplied to control how the cached function responds to unhashable arguments.  The options are:
  *  "error" (default) - Raise a `TypeError`
  *  "warning"         - Raise a `UserWarning` and call the wrapped function with the supplied arguments.
  *  "ignore"          - Just call the wrapped function with the supplied arguments.

Install
-------

Via [pip](https://pypi.python.org/pypi/fastcache):

    pip install fastcache

Manually :

    git clone https://github.com/pbrady/fastcache.git
    cd fastcache
    python setup.py install

Via [conda](http://conda.pydata.org/docs/index.html) :
  
  * build latest and greatest github version

```bash  
git clone https://github.com/pbrady/fastcache.git
conda-build fastcache
conda install --use-local fastcache
```

  * build latest released version on pypi
```bash
git clone https://github.com/conda/conda-recipes.git
conda-build conda-recipes/fastcache
conda install --use-local fastcache
```

Test
----

```python
>>> import fastcache
>>> fastcache.test()
```

Travis CI status :  [![alt text][2]][1]

[2]: https://travis-ci.org/pbrady/fastcache.svg?branch=master (Travis build status)
[1]: http://travis-ci.org/pbrady/fastcache

Tests include the official suite of tests from Python standard library for functools.lru_cache

Use
---

    >>> from fastcache import clru_cache, __version__
    >>> __version__
    '0.3.3'
    >>> @clru_cache(maxsize=325, typed=False)
    ... def fib(n):
    ...     """Terrible Fibonacci number generator."""
    ...     return n if n < 2 else fib(n-1) + fib(n-2)
    ...
    >>> fib(300)
    222232244629420445529739893461909967206666939096499764990979600
    >>> fib.cache_info()
    CacheInfo(hits=298, misses=301, maxsize=325, currsize=301)
    >>> print(fib.__doc__)
    Terrible Fibonacci number generator.
    >>> fib.cache_clear()
    >>> fib.cache_info()
    CacheInfo(hits=0, misses=0, maxsize=325, currsize=0)
    >>> fib.__wrapped__(300)
    222232244629420445529739893461909967206666939096499764990979600


Speed
-----

The speed up vs `lru_cache` provided by `functools` in 3.3 or 3.4 is 10x-30x depending on the function signature and whether one is comparing with 3.3 or 3.4.  A sample run of the benchmarking suite for 3.3 is

	>>> import sys
	>>> sys.version_info
	sys.version_info(major=3, minor=3, micro=5, releaselevel='final', serial=0)
	>>> from fastcache import benchmark
	>>> benchmark.run()
	Test Suite 1 :

	Primarily tests cost of function call, hashing and cache hits.
	Benchmark script based on
		http://bugs.python.org/file28400/lru_cache_bench.py

	function call                 speed up
	untyped(i)                       11.31, typed(i)                         31.20
	untyped("spam", i)               16.71, typed("spam", i)                 27.50
	untyped("spam", "spam", i)       14.24, typed("spam", "spam", i)         22.62
	untyped(a=i)                     13.25, typed(a=i)                       23.92
	untyped(a="spam", b=i)           10.51, typed(a="spam", b=i)             18.58
	untyped(a="spam", b="spam", c=i)  9.34, typed(a="spam", b="spam", c=i)   16.40

				 min   mean    max
	untyped    9.337 12.559 16.706
	typed     16.398 23.368 31.197


	Test Suite 2 :

	Tests millions of misses and millions of hits to quantify
	cache behavior when cache is full.

	function call                 speed up
	untyped(i, j, a="spammy")         8.94, typed(i, j, a="spammy")          14.09

A sample run of the benchmarking suite for 3.4 is

	>>> import sys
	>>> sys.version_info
	sys.version_info(major=3, minor=4, micro=1, releaselevel='final', serial=0)
	>>> from fastcache import benchmark
	>>> benchmark.run()
	Test Suite 1 :

	Primarily tests cost of function call, hashing and cache hits.
	Benchmark script based on
		http://bugs.python.org/file28400/lru_cache_bench.py

	function call                 speed up
	untyped(i)                        9.74, typed(i)                         23.31
	untyped("spam", i)               15.21, typed("spam", i)                 20.82
	untyped("spam", "spam", i)       13.35, typed("spam", "spam", i)         17.43
	untyped(a=i)                     12.27, typed(a=i)                       19.04
	untyped(a="spam", b=i)            9.81, typed(a="spam", b=i)             14.25
	untyped(a="spam", b="spam", c=i)  7.77, typed(a="spam", b="spam", c=i)   11.61

				 min   mean    max
	untyped    7.770 11.359 15.210
	typed     11.608 17.743 23.311


	Test Suite 2 :

	Tests millions of misses and millions of hits to quantify
	cache behavior when cache is full.

	function call                 speed up
	untyped(i, j, a="spammy")         8.27, typed(i, j, a="spammy")          11.18


CHANGELOG
*1.0.2*
- use pytest for testing
- Bug fix for windows compatibility

*1.0.1*
- better error checking so fastcache now plays well with signals.  There is a performance hit for this.  Next Release should handle this better.

*1.0.0*

- clru_cache now supports dynamic attributes.
- (c)lru_cache is now threadsafe via custom reentrant locks.

*0.4.3*

- Fixed bug in hash computations which resulted in `stack overflow`.  The appropriate error (RuntimeError) is now returned

*0.4.2*

- The 'state' argument to clru_cache can now be a list or a dict
- Slight performance improvemants
- Fixed compiler warnings for Python 2 builds
- Use setuptools by default.  The environment variable USE_DISTUTILS=True
  forces the use of distutils

*0.4.0*

API change:

Default behavior of fastcache is changed to raise TypeError on
unhashable arguments to be 100% consistent with stdlib.

Introduce a new argument 'unhashable' which controls how fastcache
responds to unhashable arguments:
	*'error' (default) - raise TypeError
	*'warning' - raise UserWarning and call decorated function with args
	*'ignore'  - call decorated function



fastcache/tests/test_lrucache.py
import unittest
from fastcache import lru_cache
import itertools

try:
    itertools.count(start=0, step=-1)
    count = itertools.count
except TypeError:
    def count(start=0, step=1):
        i = step-1
        for j, c in enumerate(itertools.count(start)):
            yield c + i*j

class TestLru_Cache(unittest.TestCase):
    """ Tests for lru_cache. """

    def setUp(self):

        def arg_gen(min=1, max=100, repeat=3):
            for i in range(min, max):
                for r in range(repeat):
                    for j, k in zip(range(i), count(i, -1)):
                        yield j, k

        def func(a, b):
            return a+b

        self.func = func
        self.arg_gen = arg_gen

    def test_function_attributes(self):
        """ Simple tests for attribute preservation. """

        def tfunc(a, b):
            """test function docstring."""
            return a + b
        cfunc = lru_cache()(tfunc)
        self.assertEqual(cfunc.__doc__,
                         tfunc.__doc__)

        self.assertTrue(hasattr(cfunc, 'cache_info'))
        self.assertTrue(hasattr(cfunc, 'cache_clear'))
        self.assertTrue(hasattr(cfunc, '__wrapped__'))

    def test_function_cache(self):
        """ Test that cache returns appropriate values. """

        cat_tuples = [True]

        def tfunc(a, b, c=None):
            if (cat_tuples[0] == True):
                return (a, b, c) + (c, a)
            else:
                return 2*a-10*b

        cfunc = lru_cache(maxsize=100,state=cat_tuples)(tfunc)

        for i, j in self.arg_gen(max=75, repeat=5):
            self.assertEqual(cfunc(i, j), tfunc(i, j))

        # change extra state
        cat_tuples[0] = False

        for i, j in self.arg_gen(max=75, repeat=5):
            self.assertEqual(cfunc(i, j), tfunc(i, j))

    def test_memory_leaks(self):
        """ Longer running test to check for memory leaks. """

        def tfunc(a, b, c):
            return (a-1, 2*c) + (10*b-1, a*b, a*b+c)

        cfunc = lru_cache(maxsize=2000)(tfunc)

        for i, j in self.arg_gen(max=1500, repeat=5):
            self.assertEqual(cfunc(i, j, c=i-j), tfunc(i, j, c=i-j))

    def test_warn_unhashable_args(self):
        """ Function arguments must be hashable. """

        @lru_cache(unhashable='warning')
        def f(a, b):
            return (a, ) + (b, )

        if hasattr(self, 'assertWarns'):
            with self.assertWarns(UserWarning) as cm:
                self.assertEqual(f([1], 2), f.__wrapped__([1], 2))
        else:
            #with self.assertRaises(UserWarning) as cm:
            self.assertEqual(f([1], 2), f.__wrapped__([1], 2))

    def test_ignore_unhashable_args(self):
        """ Function arguments must be hashable. """

        @lru_cache(unhashable='ignore')
        def f(a, b):
            return (a, ) + (b, )

        self.assertEqual(f([1], 2), f.__wrapped__([1], 2))

    def test_default_unhashable_args(self):
        @lru_cache()
        def f(a, b):
            return (a, ) + (b, )

        self.assertRaises(TypeError, f, [1], 2)

        @lru_cache(unhashable='error')
        def f(a, b):
            pass
        self.assertRaises(TypeError, f, [1], 2)

    def test_state_type(self):
        """ State must be a list. """

        cache = lru_cache(state=(1))
        self.assertRaises(TypeError, cache, self.func)
        cache = lru_cache(state=-1)
        self.assertRaises(TypeError, cache, self.func)

    def test_typed_False(self):
        """ Verify typed==False. """

        cfunc = lru_cache(typed=False)(self.func)
        # initialize cache with integer args
        cfunc(1,2)
        if hasattr(self, 'assertIs'):
            myAssert = self.assertIs
        else:
            myAssert = self.assertEqual
        myAssert(cfunc(1, 2), cfunc(1.0, 2))
        myAssert(cfunc(1, 2), cfunc(1, 2.0))
        # test keywords
        cfunc(1,b=2)
        myAssert(cfunc(1,b=2), cfunc(1.0,b=2))
        myAssert(cfunc(1,b=2), cfunc(1,b=2.0))

    def test_typed_True(self):
        """ Verify typed==True. """

        cfunc = lru_cache(typed=True)(self.func)
        if hasattr(self, 'assertIsNot'):
            myAssert = self.assertIsNot
        else:
            myAssert = self.assertEqual
        myAssert(cfunc(1, 2), cfunc(1.0, 2))
        myAssert(cfunc(1, 2), cfunc(1, 2.0))
        # test keywords
        myAssert(cfunc(1,b=2), cfunc(1.0,b=2))
        myAssert(cfunc(1,b=2), cfunc(1,b=2.0))


fastcache/tests/test_clrucache.py
import pytest
import fastcache
import itertools

try:
    itertools.count(start=0, step=-1)
    count = itertools.count
except TypeError:
    def count(start=0, step=1):
        i = step-1
        for j, c in enumerate(itertools.count(start)):
            yield c + i*j

def arg_gen(min=1, max=100, repeat=3):
    for i in range(min, max):
        for r in range(repeat):
            for j, k in zip(range(i), count(i, -1)):
                yield j, k

@pytest.fixture(scope='module', params=[fastcache.clru_cache,
                                        fastcache.lru_cache])
def cache(request):
    param = request.param
    return param


def test_function_attributes(cache):
    """ Simple tests for attribute preservation. """

    def tfunc(a, b):
        """test function docstring."""
        return a + b
    cfunc = cache()(tfunc)
    assert cfunc.__doc__ == tfunc.__doc__
    assert hasattr(cfunc, 'cache_info')
    assert hasattr(cfunc, 'cache_clear')
    assert hasattr(cfunc, '__wrapped__')


def test_function_cache(cache):
    """ Test that cache returns appropriate values. """

    cat_tuples = [True]

    def tfunc(a, b, c=None):
        if (cat_tuples[0] == True):
            return (a, b, c) + (c, a)
        else:
            return 2*a-10*b

    cfunc = cache(maxsize=100, state=cat_tuples)(tfunc)

    for i, j in arg_gen(max=75, repeat=5):
        assert cfunc(i, j) == tfunc(i, j)

    # change extra state
    cat_tuples[0] = False

    for i, j in arg_gen(max=75, repeat=5):
        assert cfunc(i, j) == tfunc(i, j)

    # test dict state
    d = {}
    cfunc = cache(maxsize=100, state=d)(tfunc)
    cfunc(1, 2)
    assert cfunc.cache_info().misses == 1
    d['a'] = 42
    cfunc(1, 2)
    assert cfunc.cache_info().misses == 2
    cfunc(1, 2)
    assert cfunc.cache_info().misses == 2
    assert cfunc.cache_info().hits == 1
    d.clear()
    cfunc(1, 2)
    assert cfunc.cache_info().misses == 2
    assert cfunc.cache_info().hits == 2
    d['a'] = 44
    cfunc(1, 2)
    assert cfunc.cache_info().misses == 3

def test_memory_leaks(cache):
    """ Longer running test to check for memory leaks. """

    def tfunc(a, b, c):
        return (a-1, 2*c) + (10*b-1, a*b, a*b+c)

    cfunc = cache(maxsize=2000)(tfunc)

    for i, j in arg_gen(max=1500, repeat=5):
        assert cfunc(i, j, c=i-j) == tfunc(i, j, c=i-j)

def test_warn_unhashable_args(cache, recwarn):
    """ Function arguments must be hashable. """

    @cache(unhashable='warning')
    def f(a, b):
        return (a, ) + (b, )

    assert f([1], 2) == f.__wrapped__([1], 2)
    w = recwarn.pop(UserWarning)
    assert issubclass(w.category, UserWarning)
    assert "Unhashable arguments cannot be cached" in str(w.message)
    assert w.filename
    assert w.lineno


def test_ignore_unhashable_args(cache):
    """ Function arguments must be hashable. """

    @cache(unhashable='ignore')
    def f(a, b):
        return (a, ) + (b, )

    assert f([1], 2) == f.__wrapped__([1], 2)

def test_default_unhashable_args(cache):
    @cache()
    def f(a, b):
        return (a, ) + (b, )

    with pytest.raises(TypeError):
        f([1], 2)

    @cache(unhashable='error')
    def f(a, b):
        pass
    with pytest.raises(TypeError):
        f([1], 2)

def test_state_type(cache):
    """ State must be a list or dict. """
    f = lambda x : x
    with pytest.raises(TypeError):
        cache(state=(1, ))(f)
    with pytest.raises(TypeError):
        cache(state=-1)(f)

def test_typed_False(cache):
    """ Verify typed==False. """

    @cache(typed=False)
    def cfunc(a, b):
        return a+b

    # initialize cache with integer args
    cfunc(1, 2)
    assert cfunc(1, 2) is cfunc(1.0, 2)
    assert cfunc(1, 2) is cfunc(1, 2.0)
    # test keywords
    cfunc(1, b=2)
    assert cfunc(1,b=2) is cfunc(1.0,b=2)
    assert cfunc(1,b=2) is cfunc(1,b=2.0)

def test_typed_True(cache):
    """ Verify typed==True. """

    @cache(typed=True)
    def cfunc(a, b):
        return a+b

    assert cfunc(1, 2) is not cfunc(1.0, 2)
    assert cfunc(1, 2) is not cfunc(1, 2.0)
    # test keywords
    assert cfunc(1,b=2) is not cfunc(1.0,b=2)
    assert cfunc(1,b=2) is not cfunc(1,b=2.0)

def test_dynamic_attribute(cache):
    f = lambda x : x
    cfunc = cache()(f)
    cfunc.new_attr = 5
    assert cfunc.new_attr == 5


fastcache/tests/test_functools.py
# Copied from python src Python-3.4.0/Lib/test/test_functools.py

import abc
import collections
from itertools import permutations
import pickle
from random import choice
import sys
import unittest
import fastcache
import functools

class TestLRU(unittest.TestCase):

    def test_lru(self):
        def orig(x, y):
            return 3 * x + y
        f = fastcache.clru_cache(maxsize=20)(orig)
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(maxsize, 20)
        self.assertEqual(currsize, 0)
        self.assertEqual(hits, 0)
        self.assertEqual(misses, 0)

        domain = range(5)
        for i in range(1000):
            x, y = choice(domain), choice(domain)
            actual = f(x, y)
            expected = orig(x, y)
            self.assertEqual(actual, expected)
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertTrue(hits > misses)
        self.assertEqual(hits + misses, 1000)
        self.assertEqual(currsize, 20)

        f.cache_clear()   # test clearing
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(hits, 0)
        self.assertEqual(misses, 0)
        self.assertEqual(currsize, 0)
        f(x, y)
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(hits, 0)
        self.assertEqual(misses, 1)
        self.assertEqual(currsize, 1)

        # Test bypassing the cache
        self.assertIs(f.__wrapped__, orig)
        f.__wrapped__(x, y)
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(hits, 0)
        self.assertEqual(misses, 1)
        self.assertEqual(currsize, 1)

        # test size zero (which means "never-cache")
        @fastcache.clru_cache(0)
        def f():
            #nonlocal f_cnt
            f_cnt[0] += 1
            return 20
        self.assertEqual(f.cache_info().maxsize, 0)
        f_cnt = [0]
        for i in range(5):
            self.assertEqual(f(), 20)
        self.assertEqual(f_cnt, [5])
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(hits, 0)
        self.assertEqual(misses, 5)
        self.assertEqual(currsize, 0)

        # test size one
        @fastcache.clru_cache(1)
        def f():
            #nonlocal f_cnt
            f_cnt[0] += 1
            return 20
        self.assertEqual(f.cache_info().maxsize, 1)
        f_cnt[0] = 0
        for i in range(5):
            self.assertEqual(f(), 20)
        self.assertEqual(f_cnt, [1])
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(hits, 4)
        self.assertEqual(misses, 1)
        self.assertEqual(currsize, 1)

        # test size two
        @fastcache.clru_cache(2)
        def f(x):
            #nonlocal f_cnt
            f_cnt[0] += 1
            return x*10
        self.assertEqual(f.cache_info().maxsize, 2)
        f_cnt[0] = 0
        for x in 7, 9, 7, 9, 7, 9, 8, 8, 8, 9, 9, 9, 8, 8, 8, 7:
            #    *  *              *                          *
            self.assertEqual(f(x), x*10)
        self.assertEqual(f_cnt, [4])
        hits, misses, maxsize, currsize = f.cache_info()
        self.assertEqual(hits, 12)
        self.assertEqual(misses, 4)
        self.assertEqual(currsize, 2)

    def test_lru_with_maxsize_none(self):
        @fastcache.clru_cache(maxsize=None)
        def fib(n):
            if n < 2:
                return n
            return fib(n-1) + fib(n-2)
        self.assertEqual([fib(n) for n in range(16)],
            [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610])
        self.assertEqual(fib.cache_info(),
            functools._CacheInfo(hits=28, misses=16, maxsize=None, currsize=16))
        fib.cache_clear()
        self.assertEqual(fib.cache_info(),
            functools._CacheInfo(hits=0, misses=0, maxsize=None, currsize=0))

    def test_lru_with_exceptions(self):
        # Verify that user_function exceptions get passed through without
        # creating a hard-to-read chained exception.
        # http://bugs.python.org/issue13177
        for maxsize in (None, 128):
            @fastcache.clru_cache(maxsize)
            def func(i):
                return 'abc'[i]
            self.assertEqual(func(0), 'a')
            with self.assertRaises(IndexError) as cm:
                func(15)
            self.assertIsNone(cm.exception.__context__)
            # Verify that the previous exception did not result in a cached entry
            with self.assertRaises(IndexError):
                func(15)

    def test_lru_with_types(self):
        for maxsize in (None, 128):
            @fastcache.clru_cache(maxsize=maxsize, typed=True)
            def square(x):
                return x * x
            self.assertEqual(square(3), 9)
            self.assertEqual(type(square(3)), type(9))
            self.assertEqual(square(3.0), 9.0)
            self.assertEqual(type(square(3.0)), type(9.0))
            self.assertEqual(square(x=3), 9)
            self.assertEqual(type(square(x=3)), type(9))
            self.assertEqual(square(x=3.0), 9.0)
            self.assertEqual(type(square(x=3.0)), type(9.0))
            self.assertEqual(square.cache_info().hits, 4)
            self.assertEqual(square.cache_info().misses, 4)

    def test_lru_with_keyword_args(self):
        @fastcache.clru_cache()
        def fib(n):
            if n < 2:
                return n
            return fib(n=n-1) + fib(n=n-2)
        self.assertEqual(
            [fib(n=number) for number in range(16)],
            [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610]
        )
        self.assertEqual(fib.cache_info(),
            functools._CacheInfo(hits=28, misses=16, maxsize=128, currsize=16))
        fib.cache_clear()
        self.assertEqual(fib.cache_info(),
            functools._CacheInfo(hits=0, misses=0, maxsize=128, currsize=0))

    def test_lru_with_keyword_args_maxsize_none(self):
        @fastcache.clru_cache(maxsize=None)
        def fib(n):
            if n < 2:
                return n
            return fib(n=n-1) + fib(n=n-2)
        self.assertEqual([fib(n=number) for number in range(16)],
            [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610])
        self.assertEqual(fib.cache_info(),
            functools._CacheInfo(hits=28, misses=16, maxsize=None, currsize=16))
        fib.cache_clear()
        self.assertEqual(fib.cache_info(),
            functools._CacheInfo(hits=0, misses=0, maxsize=None, currsize=0))

    def test_need_for_rlock(self):
        # This will deadlock on an LRU cache that uses a regular lock

        @fastcache.clru_cache(maxsize=10)
        def test_func(x):
            'Used to demonstrate a reentrant lru_cache call within a single thread'
            return x

        class DoubleEq:
            'Demonstrate a reentrant lru_cache call within a single thread'
            def __init__(self, x):
                self.x = x
            def __hash__(self):
                return self.x
            def __eq__(self, other):
                if self.x == 2:
                    test_func(DoubleEq(1))
                return self.x == other.x

        test_func(DoubleEq(1))                      # Load the cache
        test_func(DoubleEq(2))                      # Load the cache
        self.assertEqual(test_func(DoubleEq(2)),    # Trigger a re-entrant __eq__ call
                         DoubleEq(2))               # Verify the correct return value


fastcache/__init__.py
""" C implementation of LRU caching.

Provides 2 LRU caching function decorators:

clru_cache - built-in (faster)
           >>> from fastcache import clru_cache
           >>> @clru_cache(maxsize=128,typed=False)
           ... def f(a, b):
           ...     return (a, ) + (b, )
           ...
           >>> type(f)
           >>> <class 'fastcache.clru_cache'>

lru_cache  - python wrapper around clru_cache (slower)
           >>> from fastcache import lru_cache
           >>> @lru_cache(maxsize=128,typed=False)
           ... def f(a, b):
           ...     return (a, ) + (b, )
           ...
           >>> type(f)
           >>> <class 'function'>
"""

__version__ = "1.0.2"


from ._lrucache import clru_cache
from functools import update_wrapper

def lru_cache(maxsize=128, typed=False, state=None, unhashable='error'):
    """Least-recently-used cache decorator.

    If *maxsize* is set to None, the LRU features are disabled and
    the cache can grow without bound.

    If *typed* is True, arguments of different types will be cached
    separately. For example, f(3.0) and f(3) will be treated as distinct
    calls with distinct results.

    If *state* is a list or dict, the items will be incorporated into
    argument hash.

    The result of calling the cached function with unhashable (mutable)
    arguments depends on the value of *unhashable*:

        If *unhashable* is 'error', a TypeError will be raised.

        If *unhashable* is 'warning', a UserWarning will be raised, and
        the wrapped function will be called with the supplied arguments.
        A miss will be recorded in the cache statistics.

        If *unhashable* is 'ignore', the wrapped function will be called
        with the supplied arguments. A miss will will be recorded in
        the cache statistics.

    View the cache statistics named tuple (hits, misses, maxsize, currsize)
    with f.cache_info().  Clear the cache and statistics with
    f.cache_clear(). Access the underlying function with f.__wrapped__.

    See:  http://en.wikipedia.org/wiki/Cache_algorithms#Least_Recently_Used

    """
    def func_wrapper(func):
        _cached_func = clru_cache(maxsize, typed, state, unhashable)(func)

        def wrapper(*args, **kwargs):
            return _cached_func(*args, **kwargs)

        wrapper.__wrapped__ = func
        wrapper.cache_info = _cached_func.cache_info
        wrapper.cache_clear = _cached_func.cache_clear

        return update_wrapper(wrapper,func)

    return func_wrapper

def test(*args):
    import pytest, os
    return not pytest.main([os.path.dirname(os.path.abspath(__file__))] +
                           list(args))


scripts/threadsafety.py
from __future__ import division

"""The Python interpreter may switch between threads inbetween bytecode
execution.  Bytecode execution in fastcache may occur during:
(1) Calls to make_key which will call the __hash__ methods of the args and
(2) `PyDict_Get(Set)Item` calls rely on Python comparisons (i.e, __eq__)
    to determine if a match has been found

A good test for threadsafety is then to cache a function which takes user
defined Python objects that have __hash__ and __eq__ methods which live in
Python land rather built-in land.

The test should not only ensure that the correct result is acheived (and no
segfaults) but also assess memory leaks.

The thread switching interval can be altered using sys.setswitchinterval.
"""

class PythonInt:
    """ Wrapper for an integer with python versions of __eq__ and __hash__."""

    def __init__(self, val):
        self.value = val

    def __hash__(self):
        return hash(self.value)

    def __eq__(self, other):
        # only compare with other instances of PythonInt
        if not isinstance(other, PythonInt):
            raise TypeError("PythonInt cannot be compared to %s" % type(other))
        return self.value == other.value

from fastcache import clru_cache
#from functools import lru_cache as clru_cache
from random import randint

CACHE_SIZE=301
FIB=CACHE_SIZE-1
RAND_MIN, RAND_MAX = 1, 10

@clru_cache(maxsize=CACHE_SIZE, typed=False)
def fib(n):
    """Terrible Fibonacci number generator."""
    v = n.value
    return v if v < 2 else fib(PythonInt(v-1)) + fib(PythonInt(v-2))

# establish correct result from single threaded exectution
RESULT = fib(PythonInt(FIB))

def run_fib_with_clear(r):
    """ Run Fibonacci generator r times. """
    for i in range(r):
        if randint(RAND_MIN, RAND_MAX) == RAND_MIN:
            fib.cache_clear()
        res = fib(PythonInt(FIB))
        if RESULT != res:
            raise ValueError("Expected %d, Got %d" % (RESULT, res))

def run_fib_with_stats(r):
    """ Run Fibonacci generator r times. """
    for i in range(r):
        res = fib(PythonInt(FIB))
        if RESULT != res:
            raise ValueError("Expected %d, Got %d" % (RESULT, res))

from threading import Thread
try:
    from sys import setswitchinterval as setinterval
except ImportError:
    from sys import setcheckinterval
    def setinterval(i):
        return setcheckinterval(int(i))


def run_threads(threads):
    for t in threads:
        t.start()
    for t in threads:
        t.join()

def run_test(n, r, i):
    """ Run thread safety test with n threads r times using interval i. """
    setinterval(i)
    threads = [Thread(target=run_fib_with_clear, args=(r, )) for _ in range(n)]
    run_threads(threads)

def run_test2(n, r, i):
    """ Run thread safety test to make sure the cache statistics
    are correct."""
    fib.cache_clear()
    setinterval(i)
    threads = [Thread(target=run_fib_with_stats, args=(r, )) for _ in range(n)]
    run_threads(threads)

    hits, misses, maxsize, currsize = fib.cache_info()
    if misses != CACHE_SIZE:
        raise ValueError("Expected %d misses, Got %d" %
                         (CACHE_SIZE, misses))
    if maxsize != CACHE_SIZE:
        raise ValueError("Expected %d maxsize, Got %d" %
                         (CACHE_SIZE, maxsize))
    if currsize != CACHE_SIZE:
        raise ValueError("Expected %d currsize, Got %d" %
                         (CACHE_SIZE, currsize))

import argparse

def main():
    parser = argparse.ArgumentParser(description='Run threadsafety test.')
    parser.add_argument('-n,--numthreads',
                        type=int,
                        default=2,
                        dest='n',
                        help='Number of threads.')
    parser.add_argument('-r,--repeat',
                        type=int,
                        default=5000,
                        dest='r',
                        help='Number of times to repeat test.  Larger numbers '+
                        'will make it easier to spot memory leaks.')
    parser.add_argument('-i,--interval',
                        type=float,
                        default=1e-6,
                        dest='i',
                        help='Time in seconds for sys.setswitchinterval.')

    run_test(**dict(vars(parser.parse_args())))
    run_test2(**dict(vars(parser.parse_args())))


if __name__ == "__main__":
    main()


fastcache/benchmark.py
""" Benchmark against functools.lru_cache.

    Benchmark script from http://bugs.python.org/file28400/lru_cache_bench.py
    with a few modifications.

    Not available for Py < 3.3.
"""
from __future__ import print_function

import sys

if sys.version_info[:2] >= (3, 3):

    import functools
    import fastcache
    import timeit
    from itertools import count

    def _untyped(*args, **kwargs):
        pass

    def _typed(*args, **kwargs):
        pass

    _py_untyped = functools.lru_cache(maxsize=100)(_untyped)
    _c_untyped =  fastcache.clru_cache(maxsize=100)(_untyped)

    _py_typed = functools.lru_cache(maxsize=100, typed=True)(_typed)
    _c_typed =  fastcache.clru_cache(maxsize=100, typed=True)(_typed)

    def _arg_gen(min=1, max=100, repeat=3):
        for i in range(min, max):
            for r in range(repeat):
                for j, k in zip(range(i), count(i, -1)): 
                    yield j, k

    def _print_speedup(results):
        print('')
        print('{:9s} {:>6s} {:>6s} {:>6s}'.format('','min', 'mean', 'max'))
        def print_stats(name,off0, off1):
            arr = [py[0]/c[0] for py, c in zip(results[off0::4],
                                              results[off1::4])]
            print('{:9s} {:6.3f} {:6.3f} {:6.3f}'.format(name,
                                                         min(arr),
                                                         sum(arr)/len(arr),
                                                         max(arr)))
        print_stats('untyped', 0, 1)
        print_stats('typed', 2, 3)

    def _print_single_speedup(res=None, init=False):
        if init:
            print('{:29s} {:>8s}'.format('function call', 'speed up'))
        else:
            print('{:32s} {:5.2f}'.format(res[0][1].split('_')[-1],
                                          res[0][0]/res[1][0]), end = ', ')
            print('{:32s} {:5.2f}'.format(res[2][1].split('_')[-1],
                                          res[2][0]/res[3][0]))
    def run():

        print("Test Suite 1 : ", end='\n\n')
        print("Primarily tests cost of function call, hashing and cache hits.")
        print("Benchmark script based on")
        print("    http://bugs.python.org/file28400/lru_cache_bench.py",
              end = '\n\n')

        _print_single_speedup(init=True)

        results = []
        args = ['i', '"spam", i', '"spam", "spam", i',
                'a=i', 'a="spam", b=i', 'a="spam", b="spam", c=i']
        for a in args:
            for f in ['_py_untyped', '_c_untyped', '_py_typed', '_c_typed']:
                s = '%s(%s)' % (f, a)
                t = min(timeit.repeat('''
                for i in range(100):
                    {}
                '''.format(s),
                        setup='from fastcache.benchmark import %s' % f,
                        repeat=10, number=1000))
                results.append([t, s])
            _print_single_speedup(results[-4:])

        _print_speedup(results)

        print("\n\nTest Suite 2 :", end='\n\n')
        print("Tests millions of misses and millions of hits to quantify")
        print("cache behavior when cache is full.", end='\n\n')
        setup = "from fastcache.benchmark import {}\n" + \
                "from fastcache.benchmark import _arg_gen"

        results = []
        for f in ['_py_untyped', '_c_untyped', '_py_typed', '_c_typed']:
            s = '%s(i, j, a="spammy")' % f
            t = min(timeit.repeat('''
            for i, j in _arg_gen():
                %s
            ''' % s, setup=setup.format(f),
                                  repeat=3, number=100))
            results.append([t, s])

        _print_single_speedup(init=True)
        _print_single_speedup(results)
